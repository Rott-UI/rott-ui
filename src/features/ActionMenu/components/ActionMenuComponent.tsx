// React Imports
import React, {FC} from 'react'

// Component Imports
import {Item, Pressable, Label, Button, List, ActionMenu} from '../..'
import {ActionMenuStyles} from '../styles'
import {ActionMenuProps, ActionModel} from '../models'

// Constant Imports

// Util and Lib Imports

/**
 * Action Menu
 *
 * @param title Başlık alanı
 * @param subTitle Alt başlık alanı
 * @param data ActionModel[] formatında veri listesi
 * @param visible ActionMenu komponentin bulunduğu yerdeki Show state i
 * @returns Action Menu Renderlanır
 *
 */
export const ActionMenuComponent: FC<ActionMenuProps> = ({
  title,
  subTitle,
  data,
  itemHeight,
  maxItem,
  separatorTotalHeight,
}) => {
  const itemCount = (data?.length ?? 0) > maxItem! ? maxItem! : data?.length ?? 0

  // Listede renderlanacak Item
  const renderItem = ({
    title: renderItemTitle,
    name: renderItemName,
    action: onAction,
  }: ActionModel) => (
    <Pressable
      testID={`action-menu-data-${renderItemName}-test-id`}
      height={itemHeight}
      justifyContentCenter
      alignItemsCenter
      onPress={() => onAction()}>
      <Label fontSize='lg' variant='grey-900' fontWeight={500}>
        {renderItemTitle}
      </Label>
    </Pressable>
  )

  return (
    <Item size='full' style={ActionMenuStyles().actionMenuContainer} alignItemsCenter>
      <Item
        width={342}
        backgroundColor={COLOURS.GREY100}
        borderTopStartRadius={!(title || subTitle) ? 12 : 0}
        borderTopEndRadius={!(title || subTitle) ? 12 : 0}
        borderBottomStartRadius={12}
        borderBottomEndRadius={12}
        marginBottom={16}>
        <List
          testID='action-menu-test-id'
          style={{minHeight: 2}}
          width={342}
          height={itemHeight! * itemCount + separatorTotalHeight!}
          data={data}
          renderItem={({item}) => renderItem(item)}
          horizontal={false}
          separatorVariant='neutral-alpha-200'
          renderSeparator
          headerSeparator
          footerSeparator
          estimatedItemSize={display.normalize(itemHeight!, 'height')}
          showsVerticalScrollIndicator={false}
          scrollEnabled={(data?.length ?? 0) > maxItem!}
        />
      </Item>

      <Button
        variant='primary'
        onPress={() => ActionMenu.hideActionMenu()}
        fontSize='xl'
        size='full'
        testID='action-menu-cancel-test-id'
        marginBottom={21}>
        {translator('COMMON.CANCEL')}
      </Button>
    </Item>
  )
}
