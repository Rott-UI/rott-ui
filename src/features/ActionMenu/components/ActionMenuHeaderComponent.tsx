// React Imports
import React, {FC} from 'react'

// Util and Lib Imports
import {Item, Label} from '../..'
import {themeConfig} from '../../../providers'

interface ActionMenuHeaderProps {
  title?: Nullable<string>
  subTitle?: Nullable<string>
}

export const ActionMenuHeaderComponent: FC<ActionMenuHeaderProps> = ({title, subTitle}) => {
  return (
    <Item size='full' alignItemsCenter>
      <Item testID='action-menu-header-test-id' width={342}>
        {(title || subTitle) && (
          <Item
            justifyContentCenter
            alignItemsCenter
            size='full'
            backgroundColor={themeConfig.colors['grey-100']}
            borderTopStartRadius={12}
            borderTopEndRadius={12}
            height={54}>
            {title && (
              <Label
                fontSize='xl'
                variant='grey-900'
                fontWeight={700}
                marginBottom={subTitle ? 4 : undefined}>
                {title}
              </Label>
            )}
            {subTitle && (
              <Label fontSize='sm' variant='grey-900' fontWeight={500}>
                {subTitle}
              </Label>
            )}
          </Item>
        )}
      </Item>
    </Item>
  )
}
