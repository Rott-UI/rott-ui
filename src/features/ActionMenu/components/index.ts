export * from './ActionMenuComponent'
export * from './ActionMenuHeaderComponent'
