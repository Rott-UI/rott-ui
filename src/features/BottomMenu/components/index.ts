export * from './BottomMenu'
export * from './BottomMenuItem'
