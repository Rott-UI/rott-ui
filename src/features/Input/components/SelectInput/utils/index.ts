export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
