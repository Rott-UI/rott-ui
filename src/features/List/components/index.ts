export * from './List'
export * from './FavoriteSwipeComponent'
