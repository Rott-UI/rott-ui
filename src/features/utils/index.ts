export * from './commonUiTestExtension'
export * from './commonUiStyleProperties'
