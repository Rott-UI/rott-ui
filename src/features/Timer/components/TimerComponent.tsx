// React Imports
import React, {FC, useEffect} from 'react'

// React Native Imports
import {StyleSheet, ViewProps} from 'react-native'

// Component Imports
import {Item, Label} from '../../../components'
import {CommonUiProps} from '../../models'
import {useTimer} from '../hooks'
import {TimerStyles} from '../style'

interface TimerProps extends ViewProps, CommonUiProps {
  initialTime: number
  color?: string
}

export const TimerComponent: FC<TimerProps> = ({color, initialTime, style, ...props}) => {
  const {time, set, stop} = useTimer()

  useEffect(() => {
    set(initialTime)

    return () => {
      stop()
    }
  }, [initialTime])

  const renderTime = () => {
    const formattedTime = `${Math.floor(time / 60)}:${(time % 60).toString().padStart(2, '0')}`

    return formattedTime
  }

  return (
    <Item style={StyleSheet.flatten([TimerStyles(props).defaultTimerContainer, style])}>
      <Label fontSize='xl' fontWeight='bold' testID='timerTestId' color={color}>
        {renderTime()}
      </Label>
    </Item>
  )
}
