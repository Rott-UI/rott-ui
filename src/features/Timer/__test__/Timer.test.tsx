// React Imports
import React from 'react'

// Component Imports
import {TimerComponent} from '../components'
import {Timer} from '..'

//Util and Lib Imports
import {render, act} from '../../../utils'

describe('Timer -> Custom Component', () => {
  const timerInitialTime = 180

  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it('timer ilk render anında snapshot ile eşleşmeli', () => {
    const renderedTimer = render(<TimerComponent initialTime={timerInitialTime} />)

    expect(renderedTimer).toMatchSnapshot()
  })

  it('initial time doğru render olmalı', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)

    await act(() => Timer.countdown())

    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))
    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
  })

  it('countdown timer azalan sekilde calismali', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.countdown())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(1000))
    expect(label).toHaveProp('children', '2:59')
    await act(() => jest.advanceTimersByTime(2000))
    expect(label).toHaveProp('children', '2:57')
  })

  it('countdown timer tamamlandığında 0:00 olmalı', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.countdown())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(180 * 1000))
    expect(label).toHaveProp('children', '0:00')
  })

  it('countdown timer calisirken stop olabilmeli', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.countdown())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(10 * 1000))
    expect(label).toHaveProp('children', '2:50')
    await act(() => Timer.stop())
    expect(label).toHaveProp('children', '2:50')
  })

  it('countdown timer calisirken reset olabilmeli', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.countdown())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(10 * 1000))
    expect(label).toHaveProp('children', '2:50')
    await act(() => Timer.reset())
    expect(label).toHaveProp('children', '0:00')
  })

  it('circle timer 0:00 olabilmeli', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.circle())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(180 * 1000))
    expect(label).toHaveProp('children', '0:00')
  })

  it('circle timer 0:00 oldugunda yeniden baslamali', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.circle())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(180 * 1000))
    expect(label).toHaveProp('children', '0:00')
    rerender(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.circle())
    await act(() => jest.advanceTimersByTime(1 * 1000))
    expect(label).toHaveProp('children', '2:59')
  })

  it('circle timer calisirken stop olabilmeli', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.circle())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(10 * 1000))
    expect(label).toHaveProp('children', '2:50')
    await act(() => Timer.stop())
    expect(label).toHaveProp('children', '2:50')
  })

  it('circle timer calisirken stop olmuşken tekrar start olabilmeli', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.circle())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(10 * 1000))
    expect(label).toHaveProp('children', '2:50')
    await act(() => Timer.stop())
    expect(label).toHaveProp('children', '2:50')
    await act(() => Timer.circle())
    await act(() => jest.advanceTimersByTime(10 * 1000))
    rerender(<TimerComponent initialTime={timerInitialTime} />)
    expect(label).toHaveProp('children', '2:50')
  })

  it('circle timer calisirken reset olabilmeli', async () => {
    const {getByTestId, rerender} = render(<TimerComponent initialTime={timerInitialTime} />)
    await act(() => Timer.circle())
    await act(() => rerender(<TimerComponent initialTime={timerInitialTime} />))

    const label = getByTestId('timerTestId')

    expect(label).toHaveProp('children', '3:00')
    await act(() => jest.advanceTimersByTime(10 * 1000))
    expect(label).toHaveProp('children', '2:50')
    await act(() => Timer.reset())
    expect(label).toHaveProp('children', '0:00')
  })
})
