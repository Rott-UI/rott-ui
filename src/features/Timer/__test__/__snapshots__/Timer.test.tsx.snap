// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer -> Custom Component timer ilk render anında snapshot ile eşleşmeli 1`] = `
[
  <View
    style={
      {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "bottom": undefined,
        "flex": undefined,
        "flexDirection": "column",
        "flexGrow": undefined,
        "gap": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "left": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "maxHeight": undefined,
        "maxWidth": undefined,
        "overflow": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "paddingVertical": undefined,
        "position": undefined,
        "right": undefined,
        "top": undefined,
        "width": undefined,
        "zIndex": undefined,
      }
    }
  >
    <Text
      style={
        {
          "backgroundColor": undefined,
          "bottom": undefined,
          "color": "rgba(17, 17, 17, 1)",
          "flexGrow": undefined,
          "fontFamily": "Markpro-Bold",
          "fontSize": 29,
          "fontWeight": "bold",
          "gap": undefined,
          "height": undefined,
          "left": undefined,
          "letterSpacing": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "maxHeight": undefined,
          "maxWidth": undefined,
          "overflow": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
          "position": undefined,
          "right": undefined,
          "textAlign": "auto",
          "top": undefined,
          "width": undefined,
          "zIndex": undefined,
        }
      }
      testID="timerTestId"
    >
      3:00
    </Text>
  </View>,
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "flexDirection": "column-reverse",
          "justifyContent": "flex-start",
          "top": 30,
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>,
  <View
    onLayout={[Function]}
    pointerEvents="box-none"
    style={
      [
        {
          "alignSelf": "center",
          "elevation": 999999,
          "flex": 0,
          "maxWidth": "100%",
          "position": "absolute",
          "zIndex": 999999,
        },
        {
          "bottom": 40,
          "flexDirection": "column",
          "justifyContent": "flex-end",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        {
          "bottom": 0,
        }
      }
    >
      <RCTSafeAreaView />
    </View>
  </View>,
]
`;
